File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/NotificationHelper.java
package dev.alenajam.opendialer.feature.inCall.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import dev.alenajam.opendialer.feature.inCall.R;
import dev.alenajam.opendialer.feature.inCall.ui.InCallActivity;

public abstract class NotificationHelper {
  private static final String CHANNEL_ID_INCOMING_CALLS = "dev.alenajam.opendialer.notification_channel.incoming_calls";
  private static final String CHANNEL_ID_ONGOING_CALLS = "dev.alenajam.opendialer.notification_channel.ongoing_calls";
  private static final String CHANNEL_ID_OUTGOING_CALLS = "dev.alenajam.opendialer.notification_channel.outgoing_calls";
  private static final int NOTIFICATION_ID_CALL = 1;
  private static final String INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT = "dev.alenajam.opendialer.CALL_ACCEPT";
  private static final String INTENT_ACTION_CALL_BUTTON_CLICK_DECLINE = "dev.alenajam.opendialer.CALL_DECLINE";


  public static void setupNotificationChannels(Context context) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      createCallChannel(context, CHANNEL_ID_INCOMING_CALLS, context.getString(R.string.channel_incoming_calls), NotificationManager.IMPORTANCE_HIGH);
      createCallChannel(context, CHANNEL_ID_ONGOING_CALLS, context.getString(R.string.channel_ongoing_calls), NotificationManager.IMPORTANCE_DEFAULT);
      createCallChannel(context, CHANNEL_ID_OUTGOING_CALLS, context.getString(R.string.channel_outgoing_calls), NotificationManager.IMPORTANCE_DEFAULT);
    }
  }

  private static void createCallChannel(Context context, String channelId, String channelName, int channelImportance) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      NotificationChannel notificationChannel = new NotificationChannel(channelId, channelName, channelImportance);

      notificationChannel.setSound(null, null);

      NotificationManager notificationManager = context.getSystemService(NotificationManager.class);
      if (notificationManager != null) {
        notificationManager.createNotificationChannel(notificationChannel);
      }
    }
  }

  private static Notification notifyCall(Context context, InCallServiceImpl callService, String channelId, int priority, String notificationText) {
    Intent intent = new Intent(Intent.ACTION_MAIN, null);
    intent.setFlags(Intent.FLAG_ACTIVITY_NO_USER_ACTION | Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_BROUGHT_TO_FRONT | Intent.FLAG_ACTIVITY_NEW_DOCUMENT);
    intent.setClass(context, InCallActivity.class);
    PendingIntent pendingIntent = PendingIntent.getActivity(context, 1, intent, PendingIntent.FLAG_IMMUTABLE);

    final Notification.Builder builder;
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      builder = new Notification.Builder(context, channelId);
    } else {
      builder = new Notification.Builder(context);
    }

    builder.setPriority(priority);
    builder.setContentIntent(pendingIntent);
    builder.setFullScreenIntent(pendingIntent, true);
    builder.setSmallIcon(R.drawable.ic_notification_call);

    // TODO rework with native UI instead of RemoteViews
    /*RemoteViews remoteView = new RemoteViews(context.getPackageName(), R.layout.notification_call);
    remoteView.setImageViewBitmap(R.id.text_view, CommonUtils.textToBitmap(context, notificationText.toUpperCase(), 15, Color.BLACK));

    RemoteViews remoteViewHeadsUp = new RemoteViews(context.getPackageName(), R.layout.notification_call_heads_up);
    remoteViewHeadsUp.setImageViewBitmap(R.id.text_view, CommonUtils.textToBitmap(context, notificationText.toUpperCase(), 15, Color.BLACK));

    remoteViewHeadsUp.setImageViewBitmap(R.id.buttonLeft, CommonUtils.textToBitmap(context, context.getString(R.string.fight).toUpperCase(), 15, context.getColor(R.color.green)));
    Intent buttonLeftIntent = new Intent(context, CallButtonsListener.class);
    buttonLeftIntent.setAction(INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT);
    PendingIntent buttonLeftPendingIntent = PendingIntent.getBroadcast(context, 0, buttonLeftIntent, PendingIntent.FLAG_IMMUTABLE);
    remoteViewHeadsUp.setOnClickPendingIntent(R.id.buttonLeft, buttonLeftPendingIntent);

    remoteViewHeadsUp.setImageViewBitmap(R.id.buttonRight, CommonUtils.textToBitmap(context, context.getString(R.string.run).toUpperCase(), 15, context.getColor(R.color.red)));
    Intent buttonRightIntent = new Intent(context, CallButtonsListener.class);
    buttonRightIntent.setAction(INTENT_ACTION_CALL_BUTTON_CLICK_DECLINE);
    PendingIntent buttonRightPendingIntent = PendingIntent.getBroadcast(context, 0, buttonRightIntent, PendingIntent.FLAG_IMMUTABLE);
    remoteViewHeadsUp.setOnClickPendingIntent(R.id.buttonRight, buttonRightPendingIntent);

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      builder.setCustomHeadsUpContentView(remoteViewHeadsUp);
      builder.setCustomContentView(remoteView);
    }*/

    builder.setCategory(Notification.CATEGORY_CALL);

    builder.setVisibility(Notification.VISIBILITY_PUBLIC);

    builder.setOngoing(true);

    if (callService != null) {
      Notification notification = builder.build();
      callService.startForeground(NOTIFICATION_ID_CALL, notification);
      return notification;
    } else return null;
  }

  public static Notification notifyIncomingCall(Context context, InCallServiceImpl callService, String caller) {
    return notifyCall(context, callService, CHANNEL_ID_INCOMING_CALLS, Notification.PRIORITY_MAX, context.getString(R.string.notification_incoming_call_title, caller));
  }

  public static void notifyOutgoingCall(Context context, InCallServiceImpl callService, String caller) {
    notifyCall(context, callService, CHANNEL_ID_OUTGOING_CALLS, Notification.PRIORITY_DEFAULT, context.getString(R.string.notification_outgoing_call_title, caller));
  }

  public static void notifyOngoingCall(Context context, InCallServiceImpl callService, String caller) {
    notifyCall(context, callService, CHANNEL_ID_ONGOING_CALLS, Notification.PRIORITY_DEFAULT, context.getString(R.string.notification_ongoing_call_title, caller));
  }

  public static void notifyOnHoldCall(Context context, InCallServiceImpl callService, String caller) {
    notifyCall(context, callService, CHANNEL_ID_ONGOING_CALLS, Notification.PRIORITY_DEFAULT, context.getString(R.string.notification_on_hold_call_title, caller));
  }

  public static void removeCallNotification(InCallServiceImpl callService) {
    if (callService != null) callService.stopForeground(true);
  }

  public static void tearDown(InCallServiceImpl callService) {
    removeCallNotification(callService);
  }

  public static class CallButtonsListener extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
      if (intent.getAction() == null) return;

      CallsHandler callHandler = CallsHandler.getInstance();
      OngoingCall mainCall = callHandler.getPrimaryCall().getValue();
      if (mainCall == null) return;

      if (intent.getAction().equals(INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT)) {
        mainCall.answer();
        CallsHandler.getInstance().attemptStartActivity();
      } else {
        mainCall.hangup();
      }
    }
  }
}

File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/NotificationHelper.kt
package dev.alenajam.opendialer.feature.inCall.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Person
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import dagger.hilt.android.AndroidEntryPoint
import dev.alenajam.opendialer.feature.inCall.R
import dev.alenajam.opendialer.feature.inCall.ui.InCallActivity
import javax.inject.Inject

object NotificationHelper {
    private const val CHANNEL_ID_INCOMING_CALLS = "dev.alenajam.opendialer.notification_channel.incoming_calls"
    private const val CHANNEL_ID_ONGOING_CALLS = "dev.alenajam.opendialer.notification_channel.ongoing_calls"
    private const val CHANNEL_ID_OUTGOING_CALLS = "dev.alenajam.opendialer.notification_channel.outgoing_calls"
    private const val NOTIFICATION_ID_CALL = 1
    private const val INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT = "dev.alenajam.opendialer.CALL_ACCEPT"
    private const val INTENT_ACTION_CALL_BUTTON_CLICK_DECLINE = "dev.alenajam.opendialer.CALL_DECLINE"

    fun setupNotificationChannels(context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val nm = context.getSystemService(NotificationManager::class.java) ?: return

            val channels = listOf(
                NotificationChannel(CHANNEL_ID_INCOMING_CALLS, context.getString(R.string.channel_incoming_calls), NotificationManager.IMPORTANCE_HIGH),
                NotificationChannel(CHANNEL_ID_ONGOING_CALLS, context.getString(R.string.channel_ongoing_calls), NotificationManager.IMPORTANCE_DEFAULT),
                NotificationChannel(CHANNEL_ID_OUTGOING_CALLS, context.getString(R.string.channel_outgoing_calls), NotificationManager.IMPORTANCE_DEFAULT)
            )

            channels.forEach { channel ->
                channel.setSound(null, null)
                nm.createNotificationChannel(channel)
            }
        }
    }

    private fun notifyCall(
        context: Context,
        callService: InCallServiceImpl?,
        channelId: String,
        caller: String,
        type: CallType
    ): Notification? {
        setupNotificationChannels(context)

        val intent = Intent(context, InCallActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NO_USER_ACTION or
                    Intent.FLAG_ACTIVITY_NEW_TASK or
                    Intent.FLAG_ACTIVITY_BROUGHT_TO_FRONT or
                    Intent.FLAG_ACTIVITY_NEW_DOCUMENT
        }
        val pendingIntent = PendingIntent.getActivity(context, 1, intent, PendingIntent.FLAG_IMMUTABLE)

        val builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification.Builder(context, channelId)
        } else {
            @Suppress("DEPRECATION")
            Notification.Builder(context)
        }

        builder.setSmallIcon(R.drawable.ic_notification_call)
            .setContentIntent(pendingIntent)
            .setCategory(Notification.CATEGORY_CALL)
            .setVisibility(Notification.VISIBILITY_PUBLIC)
            .setOngoing(true)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val person = Person.Builder()
                .setName(caller)
                .setImportant(true)
                .build()
            val style = when (type) {
                CallType.INCOMING -> Notification.CallStyle.forIncomingCall(person, getDeclineIntent(context), getAcceptIntent(context))
                CallType.ONGOING -> Notification.CallStyle.forOngoingCall(person, getDeclineIntent(context))
                CallType.OUTGOING -> Notification.CallStyle.forOngoingCall(person, getDeclineIntent(context))
            }
            builder.style = style
        } else {
            builder.setContentTitle(caller)
            builder.setContentText(when (type) {
                CallType.INCOMING -> context.getString(R.string.notification_incoming_call_title, caller)
                CallType.ONGOING -> context.getString(R.string.notification_ongoing_call_title, caller)
                CallType.OUTGOING -> context.getString(R.string.notification_outgoing_call_title, caller)
            })
            if (type == CallType.INCOMING) {
                builder.addAction(Notification.Action.Builder(null, "Decline", getDeclineIntent(context)).build())
                builder.addAction(Notification.Action.Builder(null, "Answer", getAcceptIntent(context)).build())
                builder.setFullScreenIntent(pendingIntent, true)
            }
        }

        if (callService != null) {
            val notification = builder.build()
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                callService.startForeground(NOTIFICATION_ID_CALL, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL)
            } else {
                callService.startForeground(NOTIFICATION_ID_CALL, notification)
            }
            return notification
        }
        return null
    }

    private fun getAcceptIntent(context: Context): PendingIntent {
        val intent = Intent(context, CallButtonsListener::class.java).apply {
            action = INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT
        }
        return PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)
    }

    private fun getDeclineIntent(context: Context): PendingIntent {
        val intent = Intent(context, CallButtonsListener::class.java).apply {
            action = INTENT_ACTION_CALL_BUTTON_CLICK_DECLINE
        }
        return PendingIntent.getBroadcast(context, 1, intent, PendingIntent.FLAG_IMMUTABLE)
    }

    fun notifyIncomingCall(context: Context, callService: InCallServiceImpl, caller: String) =
        notifyCall(context, callService, CHANNEL_ID_INCOMING_CALLS, caller, CallType.INCOMING)

    fun notifyOutgoingCall(context: Context, callService: InCallServiceImpl, caller: String) =
        notifyCall(context, callService, CHANNEL_ID_OUTGOING_CALLS, caller, CallType.OUTGOING)

    fun notifyOngoingCall(context: Context, callService: InCallServiceImpl, caller: String) =
        notifyCall(context, callService, CHANNEL_ID_ONGOING_CALLS, caller, CallType.ONGOING)

    fun notifyOnHoldCall(context: Context, callService: InCallServiceImpl, caller: String) =
        notifyCall(context, callService, CHANNEL_ID_ONGOING_CALLS, caller, CallType.ONGOING)

    fun notifyDisconnectingCall(context: Context, callService: InCallServiceImpl, caller: String) =
        notifyCall(context, callService, CHANNEL_ID_ONGOING_CALLS, caller, CallType.ONGOING)

    fun removeCallNotification(callService: InCallServiceImpl?) {
        @Suppress("DEPRECATION")
        callService?.stopForeground(true)
    }

    fun tearDown(callService: InCallServiceImpl?) {
        removeCallNotification(callService)
    }

    private enum class CallType { INCOMING, ONGOING, OUTGOING }

    @AndroidEntryPoint
    class CallButtonsListener : BroadcastReceiver() {
        @Inject
        lateinit var callHandler: CallsHandler

        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action ?: return
            val mainCall = callHandler.displayState.value.primary ?: return

            if (action == INTENT_ACTION_CALL_BUTTON_CLICK_ACCEPT) {
                mainCall.answer()
                callHandler.attemptStartActivity()
            } else {
                mainCall.hangup()
            }
        }
    }
}
